sympy/physics/mechanics/tests/test_joint.py

```diff
@@ -0,0 +1,326 @@
+
+    #Check Linear Velocities
+    assert rod.masscenter.vel(N) == (h/4 - lB/2)*omega*A.x
+    assert plate.masscenter.vel(N) == ((h/4 - lB/2)*omega + (h/4 + lB/2)*omega)*A.x
+
+def test_pinjoint_arbitrary_axis():
+
+    theta, omega = dynamicsymbols('J_theta, J_omega', positive=True)
+
+    #When the bodies are attached though masscenters but axess are opposite.
+    N, A, P, C = _generate_body()
+    PinJoint('J', P, C, child_axis=-A.x)
+
+    assert -A.x.angle_between(N.x) == 0
+    assert -A.x.express(N) == -N.x
+    assert A.dcm(N) == Matrix([[1, 0, 0], [0, cos(theta), sin(theta)], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert A.ang_vel_in(N).magnitude() == omega
+    assert C.masscenter.pos_from(P.masscenter) == 0
+    assert C.masscenter.pos_from(P.masscenter).express(N).simplify() == 0
+    assert C.masscenter.vel(N) == 0
+
+    #When axes are different and parent joint is at masscenter but child joint is at a unit vector from
+    #child masscenter.
+    N, A, P, C = _generate_body()
+    PinJoint('J', P, C, child_axis=A.y, child_joint_pos=A.x)
+
+    assert A.y.angle_between(N.x) == 0 #Axis are aligned
+    assert A.y.express(N) == N.x
+    assert A.dcm(N) == Matrix([[0, -cos(theta), -sin(theta)], [1, 0, 0], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert A.ang_vel_in(N).express(A) == omega * A.y
+    assert A.ang_vel_in(N).magnitude() == omega
+    assert expand_mul(A.ang_vel_in(N).angle_between(A.y)) == 0
+    assert C.masscenter.vel(N) == omega*A.z
+    assert C.masscenter.pos_from(P.masscenter) == -A.x
+    assert C.masscenter.pos_from(P.masscenter).express(N).simplify() == cos(theta)*N.y + sin(theta)*N.z
+    assert C.masscenter.vel(N).angle_between(A.x) == pi/2
+
+    #Similar to previous case but wrt parent body
+    N, A, P, C = _generate_body()
+    PinJoint('J', P, C, parent_axis=N.y, parent_joint_pos=N.x)
+
+    assert N.y.angle_between(A.x) == 0 #Axis are aligned
+    assert N.y.express(A) ==  A.x
+    assert A.dcm(N) == Matrix([[0, 1, 0], [-cos(theta), 0, sin(theta)], [sin(theta), 0, cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.y
+    assert A.ang_vel_in(N).express(A) == omega* A.x
+    assert A.ang_vel_in(N).magnitude() == omega
+    assert expand_mul(A.ang_vel_in(N).angle_between(A.x)) == 0
+    assert C.masscenter.vel(N).simplify() == - omega*N.z
+    assert C.masscenter.pos_from(P.masscenter) == N.x
+
+    #Both joint pos id defined but different axes
+    N, A, P, C = _generate_body()
+    PinJoint('J', P, C, parent_joint_pos=N.x, child_joint_pos=A.x, child_axis=A.x+A.y)
+    assert expand_mul(N.x.angle_between(A.x + A.y)) == 0 #Axis are aligned
+    assert (A.x + A.y).express(N).simplify() == sqrt(2)*N.x
+    assert A.dcm(N).simplify() == Matrix([[sqrt(2)/2, -sqrt(2)*cos(theta)/2, -sqrt(2)*sin(theta)/2], \
+        [sqrt(2)/2, sqrt(2)*cos(theta)/2, sqrt(2)*sin(theta)/2], [0, -sin(theta), cos(theta)]])
```

Comment:
This direction cosine matrix matches mine above.

Reply:
Yes all other cases would match , I guess

Reply:
@moorepants Btw it's sliding joint PR